test/extensions/filters/http/cache/cache_filter_test.cc

```diff
@@ -575,6 +582,39 @@ TEST_F(CacheFilterTest, GetRequestWithBodyAndTrailers) {
   }
 }
 
+// Check that per route config is used in filter (in case if it is defined)
+TEST_F(CacheFilterTest, PerFilterConfigOverride) {
+  // Define per route config
+  const auto yaml_config = R"EOF(
+allowed_vary_headers:
+  - exact: bar
+typed_config:
+  "@type": "type.googleapis.com/envoy.extensions.cache.simple_http_cache.v3.SimpleHttpCacheConfig"
+)EOF";
+
+  envoy::extensions::filters::http::cache::v3::CacheConfig cache_per_route;
+  TestUtility::loadFromYaml(yaml_config, cache_per_route);
+
+  auto filter = makeFilter(nullptr);
+  CacheFilterConfig route_cfg(simple_cache_, cache_per_route);
+
+  ON_CALL(*decoder_callbacks_.route_, mostSpecificPerFilterConfig("envoy.filters.http.cache"))
+      .WillByDefault(Return(&route_cfg));
+
+  request_headers_.setHost("SomeHost");
+
+  // Filter cache should be null (due to we create filter with null cache - 'makeFilter(nullptr)')
+  ASSERT_TRUE(getFilterCache(filter) == nullptr);
+  EXPECT_EQ(Http::FilterHeadersStatus::StopAllIterationAndWatermark,
+            filter->decodeHeaders(request_headers_, true));
+
+  // During header decoding filter should reinit config using per route config.
+  // It means that cache should be initialized with 'simple_cache_' and allowed_vary_headers should
+  // be taken from per route config as well
+  ASSERT_TRUE(getFilterCache(filter) == simple_cache_);
+  ASSERT_TRUE(getFilterCfg(filter).proto().allowed_vary_headers_size() > 0);
```

Comment:
Similarly, you could get more readable output from this on failure, given that the configs should be identical, with
EXPECT_EQ(getFilterCfg(filter).proto().DebugString(), cache_per_route.DebugString());

In general, EXPECT_TRUE and ASSERT_TRUE shouldn't be used for conditions, they should only be used for testing boolean values - writing the condition in actual test terms will nearly always give more useful output on failure.

Ideally this one would be EXPECT_THAT(getFilterCfg(filter).proto(), ProtoEquals(cache_per_route)), but it seems ProtoEquals isn't as widely available as it should be.